File: packages/oxlint-plugin-react-doctor/src/plugin/rules/jotai/jotai-tq-use-raw-query-atom.ts
import { defineRule } from "../../utils/define-rule.js";
import { getImportedName } from "../../utils/get-imported-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import type { ScopeAnalysis } from "../../semantic/scope-analysis.js";

// HACK: jotai-tanstack-query's `atomWithQuery` returns an atom whose
// value is the full `QueryObserverResult` envelope. TanStack rebuilds
// that envelope on every observer notify, including no-op refetches.
// Subscribing directly via `useAtomValue(queryAtom)` puts every
// consumer on the broadcast path and re-renders them every notify,
// even when the field they actually read didn't change. Measured:
// 44× more commits than the equivalent `useQuery` consumer.
// Fix: derive the field once, then subscribe to the derived atom.
//   const dataAtom = atom((get) => get(queryAtom).data)
//   const data = useAtomValue(dataAtom)
// `atomWithMutation` is excluded because there's no observer envelope —
// the result IS the imperative trigger and subscribing to it is the
// documented API. `atomWithSuspenseQuery` and `atomWithInfiniteQuery`
// share the same envelope-on-every-notify shape as `atomWithQuery`.

const QUERY_ATOM_FACTORY_IMPORTED_NAMES = new Set([
  "atomWithQuery",
  "atomWithSuspenseQuery",
  "atomWithInfiniteQuery",
  "atomWithSuspenseInfiniteQuery",
]);

const SUBSCRIBING_HOOK_NAMES = new Set(["useAtomValue", "useAtom"]);

// Atoms usually live in a separate atoms module, so a name-only
// heuristic is needed for imports. The `*QueryAtom` suffix alone is
// not enough — `searchQueryAtom` is a mainstream name for a plain
// `atom('')` holding a search-query string — so the cross-file path
// additionally requires the hook result to be consumed as a
// `QueryObserverResult` envelope (see ENVELOPE_FIELD_NAMES).
const QUERY_ATOM_NAME_PATTERN = /QueryAtom$/;

// Package sources whose exports are factories/hooks, never user atoms.
const NON_ATOM_IMPORT_SOURCES = new Set(["jotai", "jotai/react", "jotai-tanstack-query"]);

const ENVELOPE_FIELD_NAMES = new Set([
  "data",
  "error",
  "status",
  "fetchStatus",
  "isLoading",
  "isError",
  "isPending",
  "isSuccess",
  "isFetching",
  "refetch",
]);

const patternDestructuresEnvelopeField = (pattern: EsTreeNode): boolean =>
  isNodeOfType(pattern, "ObjectPattern") &&
  (pattern.properties ?? []).some(
    (property) =>
      isNodeOfType(property, "Property") &&
      isNodeOfType(property.key, "Identifier") &&
      ENVELOPE_FIELD_NAMES.has(property.key.name),
  );

const isEnvelopeMemberAccess = (memberExpression: EsTreeNode): boolean =>
  isNodeOfType(memberExpression, "MemberExpression") &&
  !memberExpression.computed &&
  isNodeOfType(memberExpression.property, "Identifier") &&
  ENVELOPE_FIELD_NAMES.has(memberExpression.property.name);

const bindingIsConsumedAsEnvelope = (
  bindingIdentifier: EsTreeNode,
  scopes: ScopeAnalysis,
): boolean => {
  const symbol = scopes.symbolFor(bindingIdentifier);
  if (!symbol) return false;
  return symbol.references.some((reference) => {
    const referenceParent = reference.identifier.parent;
    if (
      isNodeOfType(referenceParent, "MemberExpression") &&
      referenceParent.object === reference.identifier
    ) {
      return isEnvelopeMemberAccess(referenceParent);
    }
    if (
      isNodeOfType(referenceParent, "VariableDeclarator") &&
      referenceParent.init === reference.identifier
    ) {
      return patternDestructuresEnvelopeField(referenceParent.id);
    }
    return false;
  });
};

// Whether the subscribe call's result is read as the query envelope:
// `useAtomValue(a).data`, `const { data } = useAtomValue(a)`,
// `const result = useAtomValue(a); result.isLoading`, and the
// `const [result] = useAtom(a)` tuple equivalents.
const isHookResultConsumedAsEnvelope = (
  callNode: EsTreeNodeOfType<"CallExpression">,
  hookName: string,
  scopes: ScopeAnalysis,
): boolean => {
  const callParent = callNode.parent;
  if (hookName === "useAtomValue") {
    if (isNodeOfType(callParent, "MemberExpression") && callParent.object === callNode) {
      return isEnvelopeMemberAccess(callParent);
    }
    if (isNodeOfType(callParent, "VariableDeclarator") && callParent.init === callNode) {
      if (patternDestructuresEnvelopeField(callParent.id)) return true;
      if (isNodeOfType(callParent.id, "Identifier")) {
        return bindingIsConsumedAsEnvelope(callParent.id, scopes);
      }
    }
    return false;
  }
  if (!isNodeOfType(callParent, "VariableDeclarator") || callParent.init !== callNode) {
    return false;
  }
  if (!isNodeOfType(callParent.id, "ArrayPattern")) return false;
  const valueElement = callParent.id.elements?.[0];
  if (!valueElement) return false;
  if (patternDestructuresEnvelopeField(valueElement)) return true;
  if (isNodeOfType(valueElement, "Identifier")) {
    return bindingIsConsumedAsEnvelope(valueElement, scopes);
  }
  return false;
};

interface SubscribeCallCandidate {
  callNode: EsTreeNodeOfType<"CallExpression">;
  hookName: string;
  atomName: string;
}

export const jotaiTqUseRawQueryAtom = defineRule({
  id: "jotai-tq-use-raw-query-atom",
  title: "Subscribing to raw query atom",
  severity: "warn",
  recommendation:
    "Derive the field you read: `const dataAtom = atom((get) => get(queryAtom).data)`. Subscribing to the whole query atom re-renders on every refetch, focus, or no-op cache hit.",
  create: (context: RuleContext) => {
    const queryAtomFactoryLocalNames = new Set<string>();
    const queryAtomBindingNames = new Set<string>();
    const importedQueryAtomNames = new Set<string>();
    const factoryCallDeclarators: EsTreeNodeOfType<"VariableDeclarator">[] = [];
    const subscribeCallCandidates: SubscribeCallCandidate[] = [];

    return {
      ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
        const source = node.source?.value;
        if (typeof source !== "string") return;
        for (const specifier of node.specifiers ?? []) {
          if (!isNodeOfType(specifier, "ImportSpecifier")) continue;
          if (!isNodeOfType(specifier.local, "Identifier")) continue;
          if (source === "jotai-tanstack-query") {
            const importedName = getImportedName(specifier);
            if (importedName && QUERY_ATOM_FACTORY_IMPORTED_NAMES.has(importedName)) {
              queryAtomFactoryLocalNames.add(specifier.local.name);
            }
            continue;
          }
          if (NON_ATOM_IMPORT_SOURCES.has(source)) continue;
          if (QUERY_ATOM_NAME_PATTERN.test(specifier.local.name)) {
            importedQueryAtomNames.add(specifier.local.name);
          }
        }
      },
      VariableDeclarator(node: EsTreeNodeOfType<"VariableDeclarator">) {
        if (!isNodeOfType(node.id, "Identifier")) return;
        const initializer: EsTreeNode | null | undefined = node.init;
        if (!isNodeOfType(initializer, "CallExpression")) return;
        if (!isNodeOfType(initializer.callee, "Identifier")) return;
        factoryCallDeclarators.push(node);
      },
      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
        if (!isNodeOfType(node.callee, "Identifier")) return;
        if (!SUBSCRIBING_HOOK_NAMES.has(node.callee.name)) return;
        const firstArgument = (node.arguments ?? [])[0];
        if (!isNodeOfType(firstArgument, "Identifier")) return;
        subscribeCallCandidates.push({
          callNode: node,
          hookName: node.callee.name,
          atomName: firstArgument.name,
        });
      },
      // Decide only after the whole file is seen — a component defined
      // ABOVE the `atomWithQuery(...)` declaration must still resolve.
      "Program:exit"() {
        for (const declarator of factoryCallDeclarators) {
          if (!isNodeOfType(declarator.id, "Identifier")) continue;
          if (!isNodeOfType(declarator.init, "CallExpression")) continue;
          if (!isNodeOfType(declarator.init.callee, "Identifier")) continue;
          if (!queryAtomFactoryLocalNames.has(declarator.init.callee.name)) continue;
          queryAtomBindingNames.add(declarator.id.name);
        }
        for (const candidate of subscribeCallCandidates) {
          const isFileLocalQueryAtom = queryAtomBindingNames.has(candidate.atomName);
          const isImportedQueryAtom =
            importedQueryAtomNames.has(candidate.atomName) &&
            isHookResultConsumedAsEnvelope(candidate.callNode, candidate.hookName, context.scopes);
          if (!isFileLocalQueryAtom && !isImportedQueryAtom) continue;
          context.report({
            node: candidate.callNode,
            message: `\`${candidate.hookName}(${candidate.atomName})\` subscribes to the whole query atom, so it re-renders your component on every refetch, focus, or no-op cache hit. Derive the field first: \`const dataAtom = atom((get) => get(${candidate.atomName}).data)\`.`,
          });
        }
      },
    };
  },
});

File: native/oxlint/rules/jotai-tq-use-raw-query-atom.rs
use oxc_ast::{
    AstKind,
    ast::{BindingPattern, Expression},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const QUERY_FACTORY_NAMES: [&str; 4] = [
    "atomWithQuery",
    "atomWithSuspenseQuery",
    "atomWithInfiniteQuery",
    "atomWithSuspenseInfiniteQuery",
];
const ENVELOPE_FIELD_NAMES: [&str; 10] = [
    "data",
    "error",
    "status",
    "fetchStatus",
    "isLoading",
    "isError",
    "isPending",
    "isSuccess",
    "isFetching",
    "refetch",
];

#[derive(Debug, Default, Clone)]
pub struct JotaiTqUseRawQueryAtom;

declare_oxc_lint!(
    /// Warn when a component subscribes to an entire Jotai query atom.
    JotaiTqUseRawQueryAtom,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Subscribing to raw query atom.",
);

impl Rule for JotaiTqUseRawQueryAtom {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(call) = node.kind() else {
            return;
        };
        let Expression::Identifier(callee) = &call.callee else {
            return;
        };
        let hook_name = callee.name.as_str();
        if !matches!(hook_name, "useAtomValue" | "useAtom") {
            return;
        }
        let Some(Expression::Identifier(atom_identifier)) = call
            .arguments
            .first()
            .and_then(oxc_ast::ast::Argument::as_expression)
        else {
            return;
        };
        let atom_name = atom_identifier.name.as_str();
        let Some(symbol_id) = ctx
            .scoping()
            .get_reference(atom_identifier.reference_id())
            .symbol_id()
        else {
            return;
        };
        let declaration = ctx.symbol_declaration(symbol_id);
        let is_file_local_query_atom = jotai_is_query_atom_declaration(declaration, ctx);
        let is_imported_query_atom = atom_name.ends_with("QueryAtom")
            && jotai_query_atom_import_is_user_owned(symbol_id, ctx)
            && jotai_hook_result_consumes_envelope(node, hook_name, ctx);
        if !is_file_local_query_atom && !is_imported_query_atom {
            return;
        }
        ctx.diagnostic(
            OxcDiagnostic::warn(format!(
                "`{hook_name}({atom_name})` subscribes to the whole query atom, so it re-renders your component on every refetch, focus, or no-op cache hit. Derive the field first: `const dataAtom = atom((get) => get({atom_name}).data)`."
            ))
            .with_label(call.span),
        );
    }
}

fn jotai_is_query_atom_declaration<'a>(declaration: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
    let AstKind::VariableDeclarator(declarator) = declaration.kind() else {
        return false;
    };
    let Some(Expression::CallExpression(factory_call)) = &declarator.init else {
        return false;
    };
    let Expression::Identifier(factory) = &factory_call.callee else {
        return false;
    };
    resolve_identifier_import(factory, ctx).is_some_and(|entry| {
        entry.module_request.name() == "jotai-tanstack-query"
            && matches!(
                &entry.import_name,
                crate::module_record::ImportImportName::Name(imported_name)
                    if QUERY_FACTORY_NAMES.contains(&imported_name.name())
            )
    })
}

fn jotai_query_atom_import_is_user_owned(
    symbol_id: oxc_semantic::SymbolId,
    ctx: &LintContext<'_>,
) -> bool {
    ctx.module_record().import_entries.iter().any(|entry| {
        !entry.is_type
            && ctx
                .scoping()
                .get_root_binding(entry.local_name.name().into())
                == Some(symbol_id)
            && matches!(
                entry.import_name,
                crate::module_record::ImportImportName::Name(_)
            )
            && !matches!(
                entry.module_request.name(),
                "jotai" | "jotai/react" | "jotai-tanstack-query"
            )
    })
}

fn jotai_hook_result_consumes_envelope<'a>(
    call_node: &AstNode<'a>,
    hook_name: &str,
    ctx: &LintContext<'a>,
) -> bool {
    let parent = ctx.nodes().parent_node(call_node.id());
    if let Some(member) = match parent.kind() {
        AstKind::StaticMemberExpression(member) => Some(member.property.name.as_str()),
        _ => None,
    } {
        return ENVELOPE_FIELD_NAMES.contains(&member);
    }
    let AstKind::VariableDeclarator(declarator) = parent.kind() else {
        return false;
    };
    let value_pattern = if hook_name == "useAtom" {
        let BindingPattern::ArrayPattern(pattern) = &declarator.id else {
            return false;
        };
        let Some(Some(element)) = pattern.elements.first() else {
            return false;
        };
        element
    } else {
        &declarator.id
    };
    if jotai_pattern_consumes_envelope(value_pattern) {
        return true;
    }
    let Some(binding) = value_pattern.get_binding_identifier() else {
        return false;
    };
    ctx.scoping()
        .get_resolved_references(binding.symbol_id())
        .any(|reference| {
            let reference_node = ctx.nodes().get_node(reference.node_id());
            let parent = ctx.nodes().parent_node(reference_node.id());
            match parent.kind() {
                AstKind::StaticMemberExpression(member) => {
                    ENVELOPE_FIELD_NAMES.contains(&member.property.name.as_str())
                }
                AstKind::VariableDeclarator(declarator) => {
                    jotai_pattern_consumes_envelope(&declarator.id)
                }
                _ => false,
            }
        })
}

fn jotai_pattern_consumes_envelope(pattern: &BindingPattern<'_>) -> bool {
    let BindingPattern::ObjectPattern(pattern) = pattern else {
        return false;
    };
    pattern.properties.iter().any(|property| {
        matches!(&property.key, oxc_ast::ast::PropertyKey::StaticIdentifier(identifier)
            if ENVELOPE_FIELD_NAMES.contains(&identifier.name.as_str()))
    })
}
